src/clippy-tracing/src/main.rs

```diff
@@ -0,0 +1,479 @@
+fn inner_instrument(_sig: &syn::Signature, suffix: &Option<String>) -> String {
+    format!(
+        "{}instrument",
+        suffix.as_ref().map_or("log_instrument::", String::as_str)
+    )
+}
+
+/// Type to return from `main` to support returning an error then handling it.
+#[repr(u8)]
+enum Exit {
+    /// Process completed successfully.
+    Ok = 0,
+    /// Process encountered an error.
+    Error = 1,
+    /// Process ran `check` action and found missing instrumentation.
+    Check = 2,
+}
+#[allow(clippy::as_conversions)]
+impl std::process::Termination for Exit {
+    fn report(self) -> ExitCode {
+        ExitCode::from(self as u8)
+    }
+}
+
+fn main() -> Exit {
+    match exec() {
+        Err(err) => {
+            eprintln!("Error: {err}");
+            Exit::Error
+        }
+        Ok(None) => Exit::Ok,
+        Ok(Some((path, line, column))) => {
+            println!(
+                "Missing instrumentation at {}:{line}:{column}.",
+                path.display()
+            );
+            Exit::Check
+        }
+    }
+}
+
+/// Error for [`exec`].
+#[derive(Debug)]
+enum ExecError {
+    /// Failed to read entry in file path.
+    Entry(walkdir::Error),
+    /// Failed to parse file path to string.
+    String,
+    /// Failed to open file.
+    File(std::io::Error),
+    /// Failed to run apply function.
+    Apply(ApplyError),
+}
+impl fmt::Display for ExecError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::Entry(entry) => write!(f, "Failed to read entry in file path: {entry}"),
+            Self::String => write!(f, "Failed to parse file path to string."),
+            Self::File(file) => write!(f, "Failed to open file: {file}"),
+            Self::Apply(apply) => write!(f, "Failed to run apply function: {apply}"),
+        }
+    }
+}
+
+impl Error for ExecError {}
+
+/// Wraps functionality from `main` to support returning an error then handling it.
+fn exec() -> Result<Option<(PathBuf, usize, usize)>, ExecError> {
+    let args = CommandLineArgs::parse();
+
+    let path = args.path.unwrap_or(PathBuf::from("."));
+    for entry_res in WalkDir::new(path).follow_links(true) {
+        let entry = entry_res.map_err(ExecError::Entry)?;
+        let entry_path = entry.into_path();
+
+        let path_str = entry_path.to_str().ok_or(ExecError::String)?;
+        // File paths must not contain any excluded strings.
+        let a = !args.exclude.iter().any(|e| path_str.contains(e));
+        // The file must not be a `build.rs` file.
+        let b = !entry_path.ends_with("build.rs");
+        // The file must be a `.rs` file.
+        let c = entry_path.extension().map_or(false, |ext| ext == "rs");
+
+        if a && b && c {
+            let file = OpenOptions::new()
+                .read(true)
+                .open(&entry_path)
+                .map_err(ExecError::File)?;
+            let res = apply(&args.action, &args.suffix, &args.cfg_attr, file, |_| {
```

Comment:
There are a large number of ways you can break it by providing a string that isn't valid syntax, most of these cases cannot be tested. I would view this in a similar light, I don't think it needs to be tested for.

Reply:
ok, would be good to have a note in the clippy-tracing/README.md that says, if clippy-tracing was run with an invalid suffix then, strip would fail and this is an expected behaviour and you'll have to discard those change so make sure to commit your changes before you run clippy-tracing

Reply:
I've added a note in the rustdoc on the argument. This will then print if a user does `clippy-tracing --help`.